### DISCUSSION TITLE
Reducing the number of {} in LaTeX

### DISCUSSION THREAD

**Discussion by akim-13** [L3MON4D3/LuaSnip#768] [Q&A] [answered]
Is it possible to write snippets that will intelligently insert curly braces in LaTeX, i.e. only when they are required? For example, `e^{2x}` needs `{}` because otherwise it would render as `(e^2) * x`, however `e^123` works just fine without `{}`. Basically, by omitting `{}` as much as possible, the expressions become a lot more readable.

**Comment by L3MON4D3** [comment, marked as answer]
Mhhh, if you write a snippet for `^`, you could do a `choiceNode` which chooses between no delimiters and `{`, `}` ([first example in this section does just that](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#restorenode)).
Maybe make it auto-triggered, if you don't have a problem with possibly triggering it accidentially.

Even closer to what you want, you could use functionNodes enclosing the text behind the `^` to automatically recognize whether the delimiters are necessary or not, and insert them accordingly. The hard part here is to figure out when the delimiters are necessary (or is it? Maybe just `"%d*"` already does it?)

**Reply by akim-13** [reply]
That's exactly what I'm looking for, thanks!

**Reply by akim-13** [reply]
For anyone interested, here is what I came up with:

```lua
local function insert_delim(args, _, delim)
    local text = args[1][1]

    -- "fb" - followed by, "fs" - forward slash
    local non_digit_fb_any_char_except_fs = string.find(text, "^%D.") and not string.find(text, "^\\")
    local digits_fb_non_digit = string.find(text, "^%d+%D")
    local fs_fb_non_spaces_fb_space = string.find(text, "^\\%S+%s")

    if non_digit_fb_any_char_except_fs or digits_fb_non_digit or fs_fb_non_spaces_fb_space then
        return delim
    -- Insert a space at the end of an expression without a delimiter
    elseif delim == "}" then
        return ' '
    else
        return ''
    end
end

local function delim_get_visual(_, parent)
    if (#parent.snippet.env.SELECT_RAW > 0) then
        return sn(nil, {
            f(insert_delim, {1}, { user_args = {"{"} }),
            i(1, parent.snippet.env.SELECT_RAW),
            f(insert_delim, {1},  { user_args = {"}"} })
        })
    else
        return sn(nil, {
            f(insert_delim, {1}, { user_args = {"{"} }),
            i(1),
            f(insert_delim, {1},  { user_args = {"}"} })
        })
    end
end

return {
    s({ name = "Superscript", trig = "([%w%)%]%}])'", wordTrig=false, regTrig = true, snippetType="autosnippet" },
        fmta(
            "<>^<>",
            {
                f( function(_, snip) return snip.captures[1] end ),
                d(1, delim_get_visual),
            }
        )
        { condition = in_mathzone }
    )
}
```
I combined inserting visual selection with "intelligent" curly braces. If you only want the latter, just return the snippet node inside `else` in `delim_get_visual` instead of the entire `if ... else`.

**Reply by L3MON4D3** [reply]
Nice!
